engine/src/execution_graph/logic_controllers/BatchAggregationProcessing.cpp

```diff
@@ -110,67 +116,83 @@ DistributeAggregateKernel::DistributeAggregateKernel(std::size_t kernel_id, cons
     std::shared_ptr<ral::cache::CacheMachine> output,
     cudaStream_t /*stream*/, const std::map<std::string, std::string>& /*args*/) {
     auto & input = inputs[0];
 
     // num_partitions = context->getTotalNodes() will do for now, but may want a function to determine this in the future.
     // If we do partition into something other than the number of nodes, then we have to use part_ids and change up more of the logic
     int num_partitions = this->context->getTotalNodes();
-
+    
     // If its an aggregation without group by we want to send all the results to the master node
     auto& self_node = ral::communication::CommunicationData::getInstance().getSelfNode();
     if (group_column_indices.size() == 0) {
-        if(this->context->isMasterNode(self_node)) {
-            bool added = this->output_.get_cache()->addToCache(std::move(input),"",false);
-            if (added) {
-                increment_node_count(self_node.id());
-            }
-        } else {
-            if (!set_empty_part_for_non_master_node){ // we want to keep in the non-master nodes something, so that the cache is not empty
-                std::unique_ptr<ral::frame::BlazingTable> empty =
-                    ral::utilities::create_empty_table(input->toBlazingTableView());
-                bool added = this->add_to_output_cache(std::move(empty), "", true);
-                set_empty_part_for_non_master_node = true;
+        try{
+            if(this->context->isMasterNode(self_node)) {
+                bool added = this->output_.get_cache()->addToCache(std::move(input),"",false);
                 if (added) {
                     increment_node_count(self_node.id());
                 }
-            }
+            } else {
+                if (!set_empty_part_for_non_master_node){ // we want to keep in the non-master nodes something, so that the cache is not empty
+                    std::unique_ptr<ral::frame::BlazingTable> empty =
+                        ral::utilities::create_empty_table(input->toBlazingTableView());
+                    bool added = this->add_to_output_cache(std::move(empty), "", true);
+                    set_empty_part_for_non_master_node = true;
+                    if (added) {
+                        increment_node_count(self_node.id());
+                    }
+                }
 
-            send_message(std::move(input),
-                true, //specific_cache
-                "", //cache_id
-                {this->context->getMasterNode().id()}); //target_id
+                send_message(std::move(input),
+                    true, //specific_cache
+                    "", //cache_id
+                    {this->context->getMasterNode().id()}); //target_id
+            }
+        }catch(rmm::bad_alloc e){
+            return {ral::execution::task_status::RETRY, std::string(e.what()), std::vector< std::unique_ptr<ral::frame::BlazingTable> > ()};
```

Comment:
how are you able to retry if you cannot return the inputs. Can we not make send_message just return this after its done?

Reply:
send_message return what when its done? It usually would just add `inputs` to the output message cache. Multiple things in that `try` move that input. So I think this is a step we cannot RETRY